MediaPlayerPrivateMediaStreamAVFObjC should ensure to have HTMLMediaElement and VideoFrame size in sync even in case of resize event

youennf opened the pull request:
#### 5be4440813a33d3ef35b4ec5d27521a692b589ad
<pre>
MediaPlayerPrivateMediaStreamAVFObjC should ensure to have HTMLMediaElement and VideoFrame size in sync even in case of resize event
<a href="https://bugs.webkit.org/show_bug.cgi?id=264803">https://bugs.webkit.org/show_bug.cgi?id=264803</a>
<a href="https://rdar.apple.com/118383141">rdar://118383141</a>

Reviewed by Eric Carlson.

When a video track size is changing, we want to make sure to fire the resize event when we receive a VideoFrame with the new size.
This ensures that creating a VideoFrame when resize event is fired will create a VideoFrame with the given size.

There are two cases where we want to keep doing what we are doing:
1. When no video frame has yet been received, the media element is in HAVE_METADATA and it is fine to fire the event more proacively when we know the size.
2. When the media stream is inactive (size is 0, 0), aka all tracks are ended, no video frame will be received but we still want to notify of the size change.
Chrome and Firefox do not follow this behavior, but we stick to it since this is preexisting behavior.

According testing, Chrome and Firefox both seem to ensure that the resize event is synchronized with creating VideoFrames with the new size.

* LayoutTests/fast/mediastream/mediastreamtrack-video-resize-event-expected.txt: Added.
* LayoutTests/fast/mediastream/mediastreamtrack-video-resize-event.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaStreamAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaStreamAVFObjC.mm:
(WebCore::m_videoLayerManager):
(WebCore::MediaPlayerPrivateMediaStreamAVFObjC::processNewVideoFrame):
(WebCore::MediaPlayerPrivateMediaStreamAVFObjC::characteristicsChanged):
(WebCore::MediaPlayerPrivateMediaStreamAVFObjC::scheduleTaskForCharacteristicsChanged):
(WebCore::MediaPlayerPrivateMediaStreamAVFObjC::videoFrameMetadata):

Canonical link: <a href="https://commits.webkit.org/270767@main">https://commits.webkit.org/270767@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/63d9047ca54a75b71744c6c1fc1a01541c0fcb84

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26328 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4938 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27587 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28425 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24097 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/26662 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6727 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2357 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/24091 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26585 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3787 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22639 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29003 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3377 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23611 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29671 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24000 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24009 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27564 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3411 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1619 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4834 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/23359 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6333 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3887 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3734 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaStreamAVFObjC.mm:
```diff
@@ -750,19 +757,26 @@ void getSupportedTypes(HashSet<String>& types) const final
+            m_intrinsicSize = intrinsicSize;
+            sizeChanged = SizeChanged::Yes;
+            if (m_playbackState == PlaybackState::None)
+                m_playbackState = PlaybackState::Paused;
```
but does this indicate that it's paused?

There could be audio playing no?

youennf replied:
m_playbackState is used to compute the display mode.
This change allows to go from `render black` to `render paused image`.
Maybe we could improve the readability of the code here.

jyavenard commented on Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaStreamAVFObjC.mm:
```diff
@@ -750,19 +757,26 @@ void getSupportedTypes(HashSet<String>& types) const final
     updateTracks();
     updateDisplayMode();
 
+    scheduleTaskForCharacteristicsChanged(sizeChanged);
```
do we need to pass sizeChanged?

why would we call scheduleTaskForCharacteristicsChanged if the size didn't change?
if the size was empty, we would fire with SizeChanged.

And if the size didn't change, there's no need to updateLayers right?

youennf replied:
This is pre-existing behaviour so I think it is best to keep it like this.
MediaPlayerPrivateMediaStreamAVFObjC::characteristicsChanged() may be called in different cases, like the removal of a track for instance, which may change the ready state (say to HaveNothing for instance).

We could probably try to make a better job at not calling scheduleTaskForCharacteristicsChanged() at all in some cases but I am not sure this is worth it.

youennf committed:
MediaPlayerPrivateMediaStreamAVFObjC should ensure to have HTMLMediaElement and VideoFrame size in sync even in case of resize event
https://bugs.webkit.org/show_bug.cgi?id=264803
rdar://118383141

Reviewed by Eric Carlson.

When a video track size is changing, we want to make sure to fire the resize event when we receive a VideoFrame with the new size.
This ensures that creating a VideoFrame when resize event is fired will create a VideoFrame with the given size.

There are two cases where we want to keep doing what we are doing:
1. When no video frame has yet been received, the media element is in HAVE_METADATA and it is fine to fire the event more proacively when we know the size.
2. When the media stream is inactive (size is 0, 0), aka all tracks are ended, no video frame will be received but we still want to notify of the size change.
Chrome and Firefox do not follow this behavior, but we stick to it since this is preexisting behavior.

According testing, Chrome and Firefox both seem to ensure that the resize event is synchronized with creating VideoFrames with the new size.

* LayoutTests/fast/mediastream/mediastreamtrack-video-resize-event-expected.txt: Added.
* LayoutTests/fast/mediastream/mediastreamtrack-video-resize-event.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaStreamAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaStreamAVFObjC.mm:
(WebCore::m_videoLayerManager):
(WebCore::MediaPlayerPrivateMediaStreamAVFObjC::processNewVideoFrame):
(WebCore::MediaPlayerPrivateMediaStreamAVFObjC::characteristicsChanged):
(WebCore::MediaPlayerPrivateMediaStreamAVFObjC::scheduleTaskForCharacteristicsChanged):
(WebCore::MediaPlayerPrivateMediaStreamAVFObjC::videoFrameMetadata):

Canonical link: https://commits.webkit.org/270767@main
